client.go

```diff
@@ -49,8 +126,17 @@ func (c *Client) client() *http.Client {
 	return c.Client
 }
 
-// do some request, but sign it before sending
+// Do some request, but sign it before sending
 func (c *Client) Do(req *http.Request) (resp *http.Response, err error) {
 	Sign(c.Auth, req)
+
+	if !c.Auth.Expiry.IsZero() {
+		if time.Now().After(c.Auth.Expiry) {
+			c.Auth.InferCredentialsFromMetadata() // this may be slow (?)
+		}
+
+		req.Header.Add(AWS_SECURITY_TOKEN_HEADER, c.Auth.Token)
```

Comment:
You could move this out of the ```IsZero``` block and add the ```X-Amz-Security-Token``` if ```c.Auth.Token``` is given. This would allow to sign with a token and only activate the refresh behavior if ```Expiry``` is given.